### PR TITLE
Reuse node-subnet from cache if exists

### PR DESCRIPTION
**Pull request by arghosh93**
**- What this PR does and why is it needed**
This PR is to return an already allocated subnet instead of creating a new one if a combination of subnet and node name already exists in the cache[1].
- May cause multiple subnet allocation per node as we currently allow same owner name for multiple subnets.

**- Special notes for reviewers**
This PR is to stop multiple allocation of node-subnet from ClusterNetwork CIDR for any node. This may occur if addition[2] of 'k8s.ovn.org/node-subnets' annotation gets delayed and meanwhile UpdateResource[3] gets executed multiple times after executing initial AddResource[3] for a node. 

Below log indicates execution of AddResource/UpdateResource(network_cluster_controller) in quick succession:

~~~
I1227 13:40:52.026089 1 node_allocator.go:371] Expected 1 subnets on node ip-10-0-134-71.us-west-2.compute.internal, found 0: []
I1227 13:40:52.042329 1 node_allocator.go:371] Expected 1 subnets on node ip-10-0-134-71.us-west-2.compute.internal, found 1: [10.128.0.0/23]
~~~

**- How to verify it**

**- Description for the changelog**

JIRA: https://issues.redhat.com/browse/OCPBUGS-25733
[1] - https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/clustermanager/node/subnet_allocator.go#L254
[2] - https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/clustermanager/node/node_allocator.go#L290
[3] -https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/clustermanager/network_cluster_controller.go

### CODE REVIEW

**Review by trozet — COMMENTED** [review]
@arghosh93 while the code change seems fine, I'm wondering how encountering this scenario ever happens. 

When HandleAddUpdateNodeEvent is called, it can only be running in one goroutine for the node (the node's name is a key so that only one goroutine will ever process this node at a time). That means that HandleAddUpdateNodeEvent 1 must have completed before HandleAddUpdateNodeEvent 2 happened. That provides 2 possible paths: HandleAddUpdateNodeEvent1 failed or succeeded. 

a. In the case where it failed, it tries to ReleaseNetwork, but this could fail and we still have something allocated, but not annotated.
b. In the case where it succeeds, the annotation must have been set. However on HandleAddUpdateNodeEvent2, I guess it is possible that when we retrieve the latest node object from the informer cache, that the kapi server has not yet sync'ed the cache. So maybe we don't see the annotation yet and accidentally allocate another subnet.

In the original bug report, it looks like there were a bunch of pods with IPs in a subnet that didnt match the currently annotated subnet on the node. I'm not sure how the scenario describedin this PR explains that.

@jcaamano wdyt?

**Comment by arghosh93** [comment]
I dont have enough evidence from log on the original bug report to claim that ReleaseNetwork failed as ovnkube-master POD was restarted after node addition. If ReleaseNetwork failed then definitely there will be an allocated network for the node without the node having annotation added to it.

I understand your hypothesis about the cache sync issue but dont have enough evidence on this front as well.

However I am pretty much sure that 2 subnets were allotted to same node. I am saying this as the subnet mentioned in other_config of logical switch is 10.179.204.0/23 and PODs were also getting IP from this subnet but node was annotated with a different subnet(10.179.206.0/23). DefaultAllow ACL was updated to match management port IP from 10.179.206.0/23 subnet it seems.

~~~
[root@c1b6ec4d19a2 ~]# ovn-nbctl --columns name,acls,other_config list logical_switch 67b426c4-8e07-48a4-be09-f6b55bae63a3
name : ip-10-177-66-41.eu-central-1.compute.internal
acls : [5e535761-412b-4215-8f0c-ea3b5071552a]
other_config : {mcast_eth_src="0a:58:0a:b3:cc:01", mcast_ip4_src="10.179.204.1", mcast_querier="true", mcast_snoop="true", subnet="10.179.204.0/23"}

[root@c1b6ec4d19a2 ~]# ovn-nbctl --columns match,priority list acl 5e535761-412b-4215-8f0c-ea3b5071552a
match : "ip4.src==10.179.206.2"
priority : 1001
~~~

This issue can happen only when subnet was allotted but node was not annotated and as a result another subnet was allotted. Hence I wanted to make sure that multiple subnet does not get allotted for same node even if node-subent annotation addition fails or gets delayed.

**Comment by jcaamano** [comment]
> @arghosh93 while the code change seems fine, I'm wondering how encountering this scenario ever happens.
> 
> When HandleAddUpdateNodeEvent is called, it can only be running in one goroutine for the node (the node's name is a key so that only one goroutine will ever process this node at a time). That means that HandleAddUpdateNodeEvent 1 must have completed before HandleAddUpdateNodeEvent 2 happened. That provides 2 possible paths: HandleAddUpdateNodeEvent1 failed or succeeded.
> 
> a. In the case where it failed, it tries to ReleaseNetwork, but this could fail and we still have something allocated, but not annotated. b. In the case where it succeeds, the annotation must have been set. However on HandleAddUpdateNodeEvent2, I guess it is possible that when we retrieve the latest node object from the informer cache, that the kapi server has not yet sync'ed the cache. So maybe we don't see the annotation yet and accidentally allocate another subnet.
> 
> In the original bug report, it looks like there were a bunch of pods with IPs in a subnet that didnt match the currently annotated subnet on the node. I'm not sure how the scenario describedin this PR explains that.
> 
> @jcaamano wdyt?

So I guess scenario B could have happened and there could have been enough time to spawn some pods? But I t would have happened in a short period of time so it is strange that it is not covered in any log. Otherwise, I still think the most probable cause is someone/somethign edited the node.

anyway, I can't think if anything against the change. I wonder if the allocator data structure would benefit from a change, given that now it needs to index by owner and not just by subnet. But I guess no harm done, not that we allocate nodes every second.

**Comment by jcaamano** [comment]
> > @arghosh93 while the code change seems fine, I'm wondering how encountering this scenario ever happens.
> > When HandleAddUpdateNodeEvent is called, it can only be running in one goroutine for the node (the node's name is a key so that only one goroutine will ever process this node at a time). That means that HandleAddUpdateNodeEvent 1 must have completed before HandleAddUpdateNodeEvent 2 happened. That provides 2 possible paths: HandleAddUpdateNodeEvent1 failed or succeeded.
> > a. In the case where it failed, it tries to ReleaseNetwork, but this could fail and we still have something allocated, but not annotated. b. In the case where it succeeds, the annotation must have been set. However on HandleAddUpdateNodeEvent2, I guess it is possible that when we retrieve the latest node object from the informer cache, that the kapi server has not yet sync'ed the cache. So maybe we don't see the annotation yet and accidentally allocate another subnet.
> > In the original bug report, it looks like there were a bunch of pods with IPs in a subnet that didnt match the currently annotated subnet on the node. I'm not sure how the scenario describedin this PR explains that.
> > @jcaamano wdyt?
> 
> So I guess scenario B could have happened and there could have been enough time to spawn some pods? But I t would have happened in a short period of time so it is strange that it is not covered in any log. Otherwise, I still think the most probable cause is someone/somethign edited the node.
> 
> anyway, I can't think if anything against the change. I wonder if the allocator data structure would benefit from a change, given that now it needs to index by owner and not just by subnet. But I guess no harm done, not that we allocate nodes every second.

On the bug, they say
```
As the PR is open for almost 3 months, is there any chance to get it reviewed? We have customers hitting this problem during upgrades and autoscaling of nodes and we would like to provide some update with some progress
```
doesn't look as something that should be complicated to reproduce and find out then.

**Comment by tssurya** [comment]
yeah when we did the troubleshooting for this bug we tried to get evidence but didn't manage to (but I know we have had two bugs https://issues.redhat.com/browse/OCPBUGS-18591 and https://issues.redhat.com/browse/OCPBUGS-25733 with same set of symptoms) => we had also asked if users were editing the nodes themselves somehow and playing with the annotations, and they said no.
I hate the fact that we don't have the evidence and the scenario we think might have happened is really a slim chance, but perhaps we merge this for now and also simultaneously ask the reporters to provide a reproducer from their side like Jaime mentioned if they are hitting this so frequently. @arghosh93

**Comment by tssurya** [comment]
@arghosh93 update here if from the 2nd bug if they are able to repro or not, if they cannot repro then we can merge this and if they can repro then we can try to RCA this accurately.

File: go-controller/pkg/clustermanager/node/subnet_allocator.go [hunk 1]
```diff
@@ -359,6 +360,18 @@ func (snr *subnetAllocatorRange) markAllocatedNetwork(owner string, network *net
 
 // allocateNetwork returns a new subnet, or nil if the range is full
 func (snr *subnetAllocatorRange) allocateNetwork(owner string) *net.IPNet {
+	// Return an already allocated subnet instead of creating a new one if a
+	// combination of subnet and node name already exists in the cache
+	for nodeSubnet, nodeName := range snr.allocMap {
+		if nodeName == owner {
+			_, subnet, err := net.ParseCIDR(nodeSubnet)
+			if err != nil {
+				klog.Errorf("Failed to parse subnet %s for node %s: %w", nodeSubnet, nodeName, err)
```
Comment (tssurya) [inline go-controller/pkg/clustermanager/node/subnet_allocator.go:369]: I think if you are only printing the error not returning it you can just use %v, why wrap it?

**Review by tssurya — COMMENTED** [review]
@arghosh93 : given both the associated bugs are closed and we don't have a clear repro here do you think we should merge this given its harmless (if so please rebase this on latest) OR should we close the PR?

**Comment by arghosh93** [comment]
@tssurya Merging this PR would definitely help. Last week I came across an issue where management port IP did not belong to node-subnets for a node. Note: Excluding hostname to protect customer data.
~~~
[arghosh@supportshell-1 03797099]$ omg get node infra03.x.x.x -oyaml|grep node-subnets
    k8s.ovn.org/node-subnets: '{"default":["10.228.2.0/23"]}'

[arghosh@supportshell-1 03797099]$ cat 0120-03797099_sosreport-infra03-2024-04-23.tar.xz/sosreport-infra03-2024-04-23/sos_commands/networking/ip_-d_address |grep ovn-k8s-mp0
6: ovn-k8s-mp0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1400 qdisc noqueue state UNKNOWN group default qlen 1000
    inet 10.229.2.2/23 brd 10.229.3.255 scope global ovn-k8s-mp0
~~~

However I still cant give proof from ovnkube log that 10.229.2.0/23 was ever allotted to infra03 node as logs have already been rotated.

File: go-controller/pkg/clustermanager/node/subnet_allocator.go [hunk 2]
```diff
@@ -359,6 +360,18 @@ func (snr *subnetAllocatorRange) markAllocatedNetwork(owner string, network *net
 
 // allocateNetwork returns a new subnet, or nil if the range is full
 func (snr *subnetAllocatorRange) allocateNetwork(owner string) *net.IPNet {
+	// Return an already allocated subnet instead of creating a new one if a
+	// combination of subnet and node name already exists in the cache
+	for nodeSubnet, nodeName := range snr.allocMap {
+		if nodeName == owner {
```
Comment (tssurya) [inline go-controller/pkg/clustermanager/node/subnet_allocator.go:366]: let's say owner = nodeA but I have a v4 and v6 range for this nodeSubnet, is that handled correctly? IIRC this subnetAllocatorRange runs a separate instances for each family?

Comment (tssurya) [reply go-controller/pkg/clustermanager/node/subnet_allocator.go:366]: I guess a nicer way to do this is to make owner a thing inside the subnetAllocatorRange but I am ok with this change as of now

Comment (arghosh93) [reply go-controller/pkg/clustermanager/node/subnet_allocator.go:366]: Yes, subnetAllocatorRange is different for different IP families.
https://github.com/arghosh93/ovn-kubernetes/blob/master/go-controller/pkg/clustermanager/node/subnet_allocator.go#L155-L185

**Review by trozet — APPROVED** [review]
I think we have all agreed that this change is fine, but we dont know if it actually is a fix to the bug that was reported. That issue has since been close due to insufficient data from the user. We will go ahead and merge this change anyway.

**Pull request merged**